buffer.c

```diff
@@ -0,0 +1,278 @@
+            if (format[1] == '!') {
+                ++format;
+                n += sizeof(long);
+                break;
+            }
+            // fall through
+
+          case 'N':  // N for big-endian 32bit unsigned integer
+          case 'V':  // V for little-endian 32bit unsigned integer
+          case 'f':  // f for native float
+          case 'e':  // e for little-endian float
+          case 'g':  // g for big-endian float
+            n += 4;
+            break;
+
+          case 'q':  // q for int64_t, q! for signed long long
+          case 'Q':  // Q for uint64_t, Q! for unsigned long long
+            if (format[1] == '!') {
+                ++format;
+                n += sizeof(LONG_LONG);
+                break;
+            }
+            // fall through
+
+          case 'd':  // d for native double
+          case 'E':  // E for little-endian double
+          case 'G':  // G for big-endian double
+            n += 8;
+            break;
+
+          case 'j':  // j for intptr_t
+          case 'J':  // J for uintptr_t
+            n += sizeof(intptr_t);
+            break;
+
+          default:
+            if (err) *err = format;
+            return -1;
+        }
+        ++format;
+    }
+
+    return n;
+}
+
+/* Return the pointer to the item located by the given indices. */
+void *
+rb_buffer_get_item_pointer(rb_buffer_t *view, ssize_t *indices)
+{
+    uint8_t *ptr = view->data;
+
+    if (view->ndim == 1) {
+        return ptr + indices[0] * view->item_size;
+    }
+
+    assert(view->shape != NULL);
+
+    int i;
+    if (view->strides == NULL) {
+        // row-major contiguous array
+        for (i = 0; i < view->ndim; ++i) {
+            ptr += indices[i] * view->shape[i] * view->item_size;
+        }
+    }
+    else if (view->sub_offsets == NULL) {
+        // flat strided array
+        for (i = 0; i < view->ndim; ++i) {
+            ptr += indices[i] * view->strides[i];
+        }
+    }
+    else {
+        // indirect strided array
+        for (i = 0; i < view->ndim; ++i) {
+            ptr += indices[i] * view->strides[i];
+            if (view->sub_offsets[i] >= 0) {
+                ptr = *(uint8_t **)ptr + view->sub_offsets[i];
+            }
+        }
+    }
+
+    return ptr;
+}
+
+static rb_buffer_protocol_entry_t *
+lookup_buffer_protocol_entry(VALUE klass) {
+    VALUE entry_obj = rb_ivar_get(klass, id_buffer_protocol);
+    while (entry_obj == Qnil) {
+        klass = rb_class_get_superclass(klass);
+
+        if (klass == rb_cBasicObject || klass == rb_cObject)
+            return NULL;
+
+        entry_obj = rb_ivar_get(klass, id_buffer_protocol);
+    }
+
+    if (! rb_typeddata_is_kind_of(entry_obj, &buffer_protocol_entry_data_type))
+        return NULL;
+
+    return (rb_buffer_protocol_entry_t *)DATA_PTR(entry_obj);
+}
+
+/* Examine whether the given object supports buffer protocol. */
+int
+rb_obj_has_buffer_protocol(VALUE obj)
```

Comment:
How about using `rb_buffer_protocol_` prefix for buffer protocol related functions?

`rb_buffer_protocol_enabled_p()`, `rb_buffer_protocol_registered_p()`, ...?